### PR TITLE
Drop `*_head` naming for structures.

### PR DESCRIPTION
**Pull request by T-256**
continued from https://github.com/ynkdir/py-win32more/pull/15 but without metaclasses!

main motivation is code readability. (e.g. in vscode go-to-definition is disabled on `*_head` annotations.)

### CODE REVIEW

**Comment by ynkdir** [comment]
The following code doesn't work.

```python
from win32more.Windows.Win32.System.Variant import VARIANT
```

**Comment by T-256** [comment]
> The following code doesn't work.
> 
> ```python
> from win32more.Windows.Win32.System.Variant import VARIANT
> ```


Thanks fixed.

Tests result:
```
E:\testwin32>python -m unittest
E
======================================================================
ERROR: win32more.Windows.Devices.PointOfService (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: win32more.Windows.Devices.PointOfService
Traceback (most recent call last):
  File "C:\Py\lib\unittest\loader.py", line 470, in _find_test_path
    package = self._get_module_from_name(name)
  File "C:\Py\lib\unittest\loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "E:\testwin32\win32more\Windows\Devices\PointOfService\__init__.py", line 4018
    _MagneticStripeReaderEncryptionAlgorithms_Meta_.None = property(get_None.__wrapped__, None)
                                                       ^
SyntaxError: invalid syntax


----------------------------------------------------------------------
Ran 1 test in 0.002s

FAILED (errors=1)
```

**Comment by ynkdir** [comment]
It doesn't work in the folloing case.

```python
# win32more.foo.bar module
class A(EasyCastStructure):
    p: POINTER(win32more.foo.bar.B)

class B(EasyCastStructure):
    a: win32more.foo.bar.A
```

then

```python
from ctypes import sizeof
from win32more.foo.bar import A
assert sizeof(A) != 0
# will fail.
```

For example.

```python
from ctypes import sizeof
from win32more.Windows.Win32.Devices.WebServicesOnDevices import WSDXML_NODE

assert sizeof(WSDXML_NODE) != 0
```

**Comment by T-256** [comment]
Nice catch!
Fixed, thanks.

File: win32more/__init__.py [hunk 1]
```diff
@@ -174,6 +212,35 @@ def easycast(obj, type_):
...
+            type_ = _patch_char_p(type_)
+
+        if isinstance(type_, BaseFuncType):
+            if not hasattr(type_, '__done_ctypes__'):
+                BaseFuncType.commit(type_)
+                type_.__done_ctypes__ = True
+            type_ = type_._fn
+        elif issubclass(type_, (EasyCastStructure, EasyCastUnion, ComPtr)):
+            type_.__commit__()
+        hints[hint] = type_
```
Comment (T-256) [inline win32more/__init__.py:240]:
```suggestion
        if not patch_return or hint == 'return':
            type_ = _patch_char_p(type_)
        if hasattr(type_, '__commit__'):
            type_.__commit__()
        hints[hint] = type_
```

File: win32more/__init__.py [hunk 2]
```diff
@@ -354,64 +413,65 @@ def wrapper(*args, **kwargs):
...
+        if not hasattr(self, '__done_ctypes__'):
+            BaseFuncType.commit(self)
+            self.__done_ctypes__ = True
+        return self._fn(*args, **kwargs)
+
+    @classmethod
+    def commit(cls, instance):
+        types = list(get_hints(instance._fn).values())
+        types = types[-1:] + types[:-1]
+        instance._fn = instance._kind(*types)
```
Comment (T-256) [inline win32more/__init__.py:431]:
```suggestion
    def __call__(self, *args, **kwargs):
        return self._fn(*args, **kwargs)

    @classmethod
    def __commit__(cls, instance):
        if isinstance(instance._fn, type(_CFuncPtr)):
            return
        types = list(get_hints(instance._fn).values())
        types = types[-1:] + types[:-1]
        instance._fn = instance._kind(*types)
```

File: win32more/__init__.py [hunk 3]
```diff
@@ -77,6 +78,19 @@ def __del__(self):
         if self and getattr(self, "_own", False):
             self.Release()
 
+    @classmethod
+    def __commit__(struct):
+        if struct.__annotations__["extends"] == 'None':
+            return
+
```
Comment (T-256) [inline win32more/__init__.py:85]:
```suggestion
        if "_hints_" in dir(struct):
            return

```

File: win32more/__init__.py [hunk 4]
```diff
@@ -354,64 +413,65 @@ def wrapper(*args, **kwargs):
...
+            raise AttributeError(
+                f"module '{self._mod}' has no attribute '{name}'"
+            ) from None
 
+        setattr(self._obj, name, prototype)
+        delattr(self._obj, f'_unused_{name}')
+
+        if issubclass(prototype, (EasyCastStructure, EasyCastUnion, ComPtr)):
+            prototype.__commit__()
+            prototype.__done_ctypes__ = True
```
Comment (T-256) [inline win32more/__init__.py:459]:
```suggestion
        if hasattr(type_, '__commit__'):
            prototype.__commit__()
```

File: win32more/__init__.py [hunk 5]
```diff
@@ -354,64 +413,65 @@ def wrapper(*args, **kwargs):
...
+def make_ready(mod: str) -> None:
+    obj = sys.modules[mod]
+
+    for name in dir(obj):
+        prototype = getattr(obj, name)
+        if (
+            isinstance(prototype, type)
+            and issubclass(prototype, (EasyCastStructure, EasyCastUnion, ComPtr))
+            and not hasattr(prototype, '__done_ctypes__')
+        ):
```
Comment (T-256) [inline win32more/__init__.py:473]:
```suggestion
        if hasattr(prototype, '__commit__'):
```

File: win32more/_winrt.py [hunk 6]
```diff
@@ -454,6 +454,9 @@ def _get_type_signature(cls) -> str:
     elif issubclass(cls, ComPtr) and "_iid_" in cls.__dict__:
         return str(cls._iid_)
     elif issubclass(cls, ComPtr):
+        if not hasattr(cls, '__done_ctypes__'):
+            cls.__commit__()
+            cls.__done_ctypes__ = True
```
Comment (T-256) [inline win32more/_winrt.py:459]:
```suggestion
        cls.__commit__()
```

**Comment by T-256** [comment]
next commit is dropping `__done_ctypes__` (it didn't test by myself, editing through github UI)

**Comment by ynkdir** [comment]
Will you complete this pull request so that unit test passes and examples works?
Otherwise, I will implement it bit differently.

**Comment by T-256** [comment]
Next commit fixes minor errors.
Improve noticeable performance by make `__commit__` one-time usable.

Tested and passed all `examples\*` with python 3.7 and 3.12
Unfortunately, `tests` folder needs rework, can you help?

**Comment by ynkdir** [comment]
I merged it with some fixing.
Thank you.

**Pull request merged**